### PR TITLE
Add Dice Loss

### PR DESCRIPTION
**Pull request by innat**
Close https://github.com/keras-team/keras-cv/issues/296

image

Ps. Tested on actual training, works fine so far for 2D and 3D cases; now needs some polish.

### CODE REVIEW

**Comment by qlzh727** [comment]
Thanks. Will take a look. (Do u still want to keep it as a draft or make it a formal PR?)

**Comment by innat** [comment]
@qlzh727 This PR can be reviewed now. Though I will add some more test cases in the `dice_test.py`, feel free to provide some initial comments. Thanks.

File: keras_cv/losses/dice.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        num_classes = tf.cast(tf.shape(y_true)[-1], y_pred.dtype)
+        return y_true * (1.0 - label_smoothing) + (label_smoothing / num_classes)
+
+    def call(self, y_true, y_pred):
+        y_pred = tf.convert_to_tensor(y_pred)
+        y_true = tf.cast(y_true, y_pred.dtype)
+        label_smoothing = tf.convert_to_tensor(self.label_smoothing, dtype=y_pred.dtype)
+
+        if self.from_logits:
+            y_pred = tf.nn.softmax(y_pred)
```
Comment (innat) [inline keras_cv/losses/dice.py:117]: @qlzh727 In [MONAI](https://github.com/Project-MONAI/MONAI), they set some arguments regarding the last layer activation function ([HERE](https://github.com/Project-MONAI/MONAI/blob/dev/monai/losses/dice.py#L48-L49)) - which helps to properly transform [logits to probabilities.](https://github.com/Project-MONAI/MONAI/blob/dev/monai/losses/dice.py#L129-L140). But keras doesn't follow this design. Instead, it provides separate loss functions (binary cross-entropy, categorical cross-entropy). 

In the above (lines 116-117), we used only softmax for `logit=True`; not for sigmoid activation. So, to maintain the keras design, what could be the fixes here?

File: keras_cv/losses/dice.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            y_true = self._smooth_labels(y_true, y_pred, label_smoothing)
+
+        if self.class_ids is not None:
+            y_true, y_pred = gather_channels(y_true, y_pred, indices=self.class_ids)
+
+        axes = (
+            tf.constant([1, 2])
+            if keras.backend.image_data_format() == "channels_last"
+            else tf.constant([2, 3])
+        )
```
Comment (innat) [inline keras_cv/losses/dice.py:129]: To support `3D`  modeling, I may need to do `tf.constant([1, 2, 3]) / tf.constant([2, 3, 4])`. But not sure how to set parameters here by maintaining the keras API design. There's no 3D version of [keras-loss](https://keras.io/api/losses/) to follow guidelines. So, need some advice here.  

An example below, (not a keras style though, asaik).

```python
if modeling == '2D':
   axes = tf.constant([1, 2])
elif modeling == '3D':
   axes = tf.constant([1, 2, 3])
else:
   raise ValueError('...')
```

Comment (innat) [reply keras_cv/losses/dice.py:129]: Thanks, @bhack. 
I was thinking about how to introduce `generalized-dice loss` here. Seeing [this](https://github.com/tensorflow/models/blob/master/official/projects/volumetric_models/losses/segmentation_losses.py#L38-L39) implementation, I think we can introduce both params if it's ok to keras-team. 

```python
def __init__(self,
      beta,
      from_logists,
      ....
      loss_type,  # 'generalized', 'adaptive', None.
      axis= (1, 2, 3)  # (1, 2, 3) for 3D, (1, 2) for 2D.
):
```

Comment (innat) [reply keras_cv/losses/dice.py]: Added the `axis` param.

File: keras_cv/losses/dice.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+                `generalized` or `adaptive`; otherwise compute original dice score.
+                Default to `None`.
+            label_smoothing: Float in [0, 1]. When > 0, label values are smoothed,
+                meaning the confidence on label values are relaxed. For example, if
+                `0.1`, use `0.1 / num_classes` for non-target labels and
+                `0.9 + 0.1 / num_classes` for target labels. Default to `0.0`.
+            per_sample: If `True`, the loss will be calculated for each sample in
+                batch and then averaged. Otherwise the loss will be calculated for
+                the whole batch. Default to `False`.
+            epsilon: Small float added to dice score to avoid dividing by zero.
```
Comment (qlzh727) [inline keras_cv/losses/dice.py:94]: I think this one is probably not needed. We can just use keras.backend.epsilon() directly.

File: keras_cv/losses/dice.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+        self.class_ids = class_ids
+
+    def _smooth_labels(self, y_true, y_pred, label_smoothing):
+        num_classes = tf.cast(tf.shape(y_true)[-1], y_pred.dtype)
+        return y_true * (1.0 - label_smoothing) + (label_smoothing / num_classes)
+
+    def call(self, y_true, y_pred):
+        y_pred = tf.convert_to_tensor(y_pred)
+        y_true = tf.cast(y_true, y_pred.dtype)
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply keras_cv/losses/dice.py:136]: Good question, I think its probably better to have 2 separate class for categorical and sparse categorical. Usually loss/metric can be created from a raw string in the model.compile. If we have a flag to config whether the losses intake one_hot or not, then user are forced to create loss instance.

File: keras_cv/losses/dice.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+            y_pred = tf.nn.softmax(y_pred)
+
+        if tf.cast(label_smoothing, dtype=tf.bool):
+            y_true = self._smooth_labels(y_true, y_pred, label_smoothing)
+
+        if self.class_ids is not None:
+            y_true, y_pred = gather_channels(y_true, y_pred, indices=self.class_ids)
+
+        # loss calculation: Fβ-score (in terms of Type I and type II erro
+        true_positive = keras.backend.sum(y_true * y_pred, axis=self.axis)
```
Comment (qlzh727) [inline keras_cv/losses/dice.py:149]: I think we are hiding most of the keras.backend functions since they are just simple wrapper of tf ops. Please replace this with tf.math ops instead.

File: keras_cv/losses/dice.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+        if self.loss_type == "generalized":
+            dice_score = self._generalized_dice_score(y_true, numerator, denominator)
+        elif self.loss_type == "adaptive":
+            dice_score = self._adaptive_dice_score(numerator, denominator)
+        else:
+            dice_score = numerator / denominator
+
+        if self.per_sample:
+            dice_score = keras.backend.mean(dice_score, axis=0)
```
Comment (qlzh727) [inline keras_cv/losses/dice.py:170]: Keras loss should return scalar tensor, so per_sample result is bit anti-pattern here.

File: keras_cv/utils/fill_utils.py [hunk 7]
```diff
@@ -80,3 +82,20 @@ def fill_rectangle(images, centers_x, centers_y, widths, heights, fill_values):
 
     images = tf.where(is_rectangle, fill_values, images)
     return images
+
+
+def gather_channels(*matrices, indices=None):
```
Comment (qlzh727) [inline keras_cv/utils/fill_utils.py:87]: This one doesn't seem have to anything to do with fill. We might just move it to the loss module for now, until there is any reuse case.

**Review by qlzh727 — CHANGES_REQUESTED** [review]
Thanks for the PR, and sorry for the late reply.

File: keras_cv/losses/dice.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_cv.utils.fill_utils import gather_channels
+
+
+class Dice(keras.losses.Loss):
```
[the remark this answers is not in the document]
Comment (innat) [reply keras_cv/losses/dice.py:21]: Done.

File: keras_cv/losses/dice.py [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+        Args:
+            beta: A float or integer coefficient for balancing the precision
+                and recall. It determines the weight of recall and precision
+                in the combined score. If `beta < 1`, precisoin will doninate;
+                if `beta > 1`, recall will dominate. Default to `1`.
+            from_logits: Whether `y_pred` is expected to be a logits tensor. By
+                default, we assume that `y_pred` encodes a probability distribution.
+                Default to `False`.
+            class_ids: An interger or a lost of intergers within `range(num_classes)`
+                to evaluate the loss. If it's `None`, all classes will beu used to
```
[the remark this answers is not in the document]
Comment (innat) [reply keras_cv/losses/dice.py]: Done.

File: keras_cv/losses/dice.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+        Args:
+            beta: A float or integer coefficient for balancing the precision
+                and recall. It determines the weight of recall and precision
+                in the combined score. If `beta < 1`, precisoin will doninate;
+                if `beta > 1`, recall will dominate. Default to `1`.
+            from_logits: Whether `y_pred` is expected to be a logits tensor. By
+                default, we assume that `y_pred` encodes a probability distribution.
+                Default to `False`.
+            class_ids: An interger or a lost of intergers within `range(num_classes)`
```
[the remark this answers is not in the document]
Comment (innat) [reply keras_cv/losses/dice.py]: Done.

File: keras_cv/losses/dice.py (hunk 5 above)
Comment (innat) [reply keras_cv/losses/dice.py]: Done.

File: keras_cv/utils/fill_utils.py (hunk 7 above)
Comment (innat) [reply keras_cv/utils/fill_utils.py]: Relocated.
Done.

File: keras_cv/losses/dice.py (hunk 3 above)
Comment (innat) [reply keras_cv/losses/dice.py:94]: This one is the smoothing parameter and sometimes users may want to set their own. WDYT? But it can be removed of course. As a default value, I use `backend.epsilon` though.

File: keras_cv/losses/dice.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        name="dice",
+        **kwargs,
+    ):
+        """Initializes `Dice` instance.
+
+        Args:
+            beta: A float or integer coefficient for balancing the precision
+                and recall. It determines the weight of recall and precision
+                in the combined score. If `beta < 1`, precisoin will doninate;
+                if `beta > 1`, recall will dominate. Default to `1`.
```
[the remark this answers is not in the document]
Comment (innat) [reply keras_cv/losses/dice.py]: Info Added. 
It should be greater than zero to any positive number.

File: keras_cv/losses/dice.py [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+                `channel_first` format respectively.
+            loss_type: An optional `str` specifying the type of the dice score to
+                compute. Compute generalized or adaptive dice score if loss type is
+                `generalized` or `adaptive`; otherwise compute original dice score.
+                Default to `None`.
+            label_smoothing: Float in [0, 1]. When > 0, label values are smoothed,
+                meaning the confidence on label values are relaxed. For example, if
+                `0.1`, use `0.1 / num_classes` for non-target labels and
+                `0.9 + 0.1 / num_classes` for target labels. Default to `0.0`.
+            per_sample: If `True`, the loss will be calculated for each sample in
```
[the remark this answers is not in the document]
Comment (innat) [reply keras_cv/losses/dice.py:92]: This param is inspired by [THIS](https://github.com/qubvel/segmentation_models/blob/94f624b7029deb463c859efbd92fa26f512b52b8/segmentation_models/losses.py#L70) implementation.

> per_image: If **True** loss is calculated for each image in batch and then averaged,
else loss is calculated for the whole batch.

The original implementation is in python, we just translate it into tf. In practice, we found that depending on the task, enabling it to `True` makes an improvement in training. So, I kept it there. WDYT?

File: keras_cv/losses/dice.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (innat) [reply keras_cv/losses/dice.py:159]: Ok. And about [this](https://github.com/keras-team/keras-cv/pull/371#discussion_r861804809), should we use another new class, `BinaryDice`?

File: keras_cv/losses/dice.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (innat) [reply keras_cv/losses/dice.py]: I did a quick look at the reference implementation but what decision should I make next for consistency? Can you elaborate on this?

The `class_id` coded [HERE](https://github.com/keras-team/keras/blob/6b3916813d761c75de61f89bde19538add1a8e2d/keras/metrics/metrics.py#L781), take only 1 class id and [slice from indices](https://github.com/keras-team/keras/blob/6b3916813d761c75de61f89bde19538add1a8e2d/keras/utils/metrics_utils.py#L613-L615), whereas, in above dice implementation, it gathers 1 and more within class indices within the valid range. Also, I'm not sure if the existing [Precision](https://github.com/keras-team/keras/blob/6b3916813d761c75de61f89bde19538add1a8e2d/keras/metrics/metrics.py#L750) (and similar others) are compatible for semantic segmentation task! I tried to run it in actual training and ended up with `InvalidArgumentError`.

File: keras_cv/losses/dice.py (hunk 1 above)
Comment (qlzh727) [reply keras_cv/losses/dice.py:163]: I think keras will probably keep the existing keras format. We don't apply any activation in the loss function. User should do this in their model instead. If user would like to apply sigmoid activation, they should change the model to include sigmoid activation, and specify a loss with from_logits=False.

File: keras_cv/losses/dice.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (qlzh727) [reply keras_cv/losses/dice.py:159]: Replied on the other thread. We should have different class for binary/categorcial/sparse target input.

**Comment by lucasdavid** [comment]
I can see `axis=[1,2]` in this and some other classes in this repository.
Don't know if this has been discussed somewhere else before, but I would like to suggest that we treated all axes as positional, except for the `batch` and `channels` axes. This way, every single method/procedure would handle multi-dimentional cases (2D, 3D etc) transparently.

This can be achieved by reshaping the 2d,3d,4d,...,nd signal into a `(batch, positions, channels)` one, and reducing over `positions`:
```py
def some_metric(y_true, y_pred, axis=None):
  if axis is not None:
    # business as usual...
    return

  y_true = squeeze(y_true, dims=3, axis=axis)      # 2d, 3d, 4d, ..., nd -> 3d
  y_pred = squeeze(y_pred, dims=3, axis=axis)    # 2d, 3d, 4d, ..., nd -> 3d

  # calculate metric result
  # ...
  return tf.reduce_mean(result, axis=1)

def squeeze(y, dims: int = 3):
  shape = tf.shape(y)

  if dims not in (2, 3):
    raise ValueError(f'Illegal value for parameter dims=`{dims}`. Can only squeeze '
                     'positional signal, resulting in a tensor with rank 2 or 3.')

  new_shape = [shape[0], -1]
  if dims == 3:  # keep channels.
    new_shape += [shape[-1]]
  return tf.reshape(y, new_shape)
```

File: keras_cv/losses/dice.py [hunk 13]
```diff
@@ -0,0 +1,517 @@
...
+            from_logits: Whether `y_pred` is expected to be a logits tensor. By
+                default, we assume that `y_pred` encodes a probability distribution.
+                Default to `False`.
+            class_ids: An interger or a list of intergers within `range(num_classes)`
+                to evaluate the loss. If it's `None`, all classes will be used to
+                calculate the loss. Default to `None`.
+            axis: An optional sequence of `int` specifying the axis to perform reduce
+                ops for raw dice score. For 2D model, it should be [1,2] or [2,3]
+                for the `channels_last` or `channels_first` format respectively. And for
+                3D mdoel, it should be [1,2,3] or [2,3,4] for the `channels_last` or
```
Comment (quantumalaviya) [inline keras_cv/losses/dice.py:146]: should be model on line 146

**Comment by LukeWood** [comment]
Thanks for your patience @innat, wanted to merge a Loss first before this.  Ill revisit this now that we have done don

**Comment by bhack** [comment]
This is the oldest open PR. What is its status?

**Comment by LukeWood** [comment]
> This is the oldest open PR. What is its status?

To be honest it simply hasn’t been prioritized due to it not being a part of one of our flagship workflows (OD w/ RetinaNet, classification training in a reusable template with SOTA augmentation, and segmentation map prediction with model TBD)

one thing to double check is that the loss reduction matches the construct set forth in the rest of the losses (see smoothL1, focal as examples).

we can revisit this, but at the moment we are focusing effort on delivering those flagship workflows.

**Comment by LukeWood** [comment]
> > This is the oldest open PR. What is its status?
> 
> To be honest it simply hasn’t been prioritized due to it not being a part of one of our flagship workflows (OD w/ RetinaNet, classification training in a reusable template with SOTA augmentation, and segmentation map prediction with model TBD)
> 
> one thing to double check is that the loss reduction matches the construct set forth in the rest of the losses (see smoothL1, focal as examples).
> 
> we can revisit this, but at the moment we are focusing effort on delivering those flagship workflows.

That explanation said @innat, feel free to rebase and re request a review.  It’s almost ready so happy to prioritize this.

**Comment by innat** [comment]
@LukeWood I may not able to work on it due to tight schedule. If it's prioritized, someone may need to take it from here. ( Same goes to [Jaccard PR](https://github.com/keras-team/keras-cv/pull/449). )

**Comment by DavidLandup0** [comment]
Can I take over this one and #449?

**Comment by DavidLandup0** [comment]
Awesome, thanks! I'll be working on these in the upcoming days then.
What's left is the documentation, test cases and implementing the suggestions from the comments, right?

**Comment by innat** [comment]
> Awesome, thanks! I'll be working on these in the upcoming days then. What's left is the documentation, test cases and implementing the suggestions from the comments, right?

Sort of.

**Comment by DavidLandup0** [comment]
@innat Could you share the training script you mentioned if you still have it?

**Comment by innat** [comment]
I am not sure which one you're referring to. Could you please more precise?  ( I'm in leave, so if I've anything that is related, I will share in upcoming days.)

**Comment by DavidLandup0** [comment]
Oh, I meant this:
> Ps. Tested on actual training, works fine so far for 2D and 3D cases; now needs some polish.

Sorry to bug you while you're on leave - enjoy your time off! :)

**Review by tanzhenyu — COMMENTED** [review]
Thanks for the PR!
A general comment -- is it possible to start with a simple one, say binary dice, test it with our existing models? This would help a LOT in speed-up the review process.

**Comment by DavidLandup0** [comment]
Of course! I'll get an example up and running as soon as I finish the test cases for the other PR

**Comment by DavidLandup0** [comment]
@tanzhenyu added an  example run in the new PR #968 based on your training script for DeepLabV3
